docs/proposals/component-registration.md

```diff
@@ -0,0 +1,294 @@
+  - Endpoints are external + internal, but only have readiness (for now)
+
+
+#### API Types
+
+The following is a draft of how the Component model might look.
+
+TODO
+
+### Kubectl (CLI)
+
+Ready Service Endpoints can be queried with the existing `kubectl get endpoints <service-name>`.
+
+Add a `kubectl get externalservicetargets` (`kubectl get est` for short) command to list the external service provider instances, their address/port/name and their last known conditions. Components can be retrived with a component specific label, like `type=component` or `component=apiserver`.
+
+
+### Storage
+
+Store ExternalServiceTargets spec/status in etcd, with an apiserver storage wrapper.
+
+
+### Status Updates
+
+Add an external-service-target-controller that regularly probes registered targets and updates their stored last known status.
+
+Add self-registration to the components that uses the new ExternalServiceTargets API (Create call) when the component starts.
+
+Add self-deregistration to the components that uses the new ExternalServiceTargets API (Remove call) when the component stops gracefully (e.g. after receiving SIGTERM).
+
+Add self-updating to the components that uses the new ExternalServiceTargets API (Update call) when the component errors/panics.
+
+Use a TCP probe (of `/healthz` or `/healthz/ping`) to check liveness, because it indicates a server is listening on the expected host/path, even if its responses are unhealthy.
+
+Update individual component `/healthz` endpoints with sub-resources to allow them to perform as an indicator of readiness, not just liveness. Readiness should include transitively checking the readiness of immediate dependencies (e.g. etcd or other databases).
+
+
+## Example
+
+The following terminal commands are an example of how the `kubectl get externalservicetargets` command might work.
+
+TODO
+
+## Alternate Naming
+
+ExternalServiceTarget is a bit verbose for a top-level API resource name. Here are some other naming options for the same concept.
+
+ExternalTarget
+[External]Endpoint (singular)
+[External]ServiceTarget
+[External]ServiceInstance
+[External]ServiceProvider
+[External]Shell (Pod without peas)
+[External]ServiceIngress
+[External]ServiceUnit
+[External]ServiceMember
+
+
+## Next Steps
+
+Since existing AddOns (e.g. kube-ui, kube-dns) already register themselves as internal services, this proposal puts them on a similar footing to externally hosted components.
+
+Services could be made to include both internal pods and external targets (in the same service).
+
+It was also suggested that this proposal overlaps somewhat with the Nodes API, and that nodes (kubelets/kube-proxy) could also be registered as ExternalServiceTargets.
+
+Because this proposal includes API changes, it's likely to be put into the experimental API group.
+
+Because of the desire for reverse compatibility, the existing `/componentstatuses` endpoint will likely be deprecated and eventually removed.
```

Comment:
We could look at reimplementing in terms of the new systems so that at least we don't have two independent codepaths.